[Jet Lee] iP

jetlfj opened the pull request:
### DukePro

> Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/jetlfj/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

jetlfj committed:
add greet, echo, exit

jetlfj committed:
add add, list

jetlfj committed:
add mark as done

jetlfj committed:
add todos, events, deadlines

jetlfj committed:
add automated text ui testing

jetlfj committed:
add handle errors

jetlfj committed:
add delete

jetlfj committed:
add use enums

jetlfj committed:
Fix formatting

The code does not follow the rules in the SE_EDU Java coding standard.

Code that follows the rules is less error-prone and is easier for
others to read and understand.

Let's update the code with the correct naming and layout.

jetlfj committed:
Add Save

Tasks are removed whenever the application is closed.

Saving and loading tasks allows users to retain the tasks they have
added.

Let's,
* save the tasks in the hard disk automatically whenever the task list
  changes
* load the data from the hard disk when the application starts up

jetlfj committed:
Add Dates and Times

The chatbot stores dates and times as a String.

Storing time as a LocalDate allows for easier comparison and formatting.

Let's
* store deadline dates as a java.time.LocalDate in task objects
* accept dates in a format such as yyyy-mm-dd and print in a different
  format such as MMM dd yyyy

jetlfj committed:
Merge branch 'branch-Level-7'

jetlfj committed:
Revert "Add Dates and Times"

This reverts commit 3e8bd19363fdff902ac733b55073949d23bdcaed.

jetlfj committed:
Add Dates and Times

The chatbot stores time as a String.

Storing time as a LocalDate allows for easier comparison and formatting.

Let's,
* store deadline dates as a java.time.LocalDate in task objects
* accept dates in a format such as yyyy-mm-dd and print in a different
  format such as MMM dd yyyy

jetlfj committed:
Merge branch 'branch-Level-8'

jetlfj committed:
Add More OOP

jetlfj committed:
Add Packages

jetlfj committed:
Fix formatting

jetlfj committed:
Merge branch 'add-gradle-support'

jetlfj committed:
Add JUnit Tests

jetlfj committed:
Add JAR

jetlfj committed:
Add JavaDoc

jetlfj committed:
Follow the Coding Standard

jetlfj committed:
Add Find

jetlfj committed:
Merge branch 'branch-A-JavaDoc'

jetlfj committed:
Merge branch 'branch-A-CodingStandard'

jetlfj committed:
Merge branch 'branch-Level-9'

jetlfj committed:
Use Checkstyle

jetlfj committed:
Add GUI

jetlfj committed:
Use Varargs

jetlfj committed:
Merge branch 'branch-A-CheckStyle'

jetlfj committed:
Merge branch 'branch-Level-10'

jetlfj committed:
Merge branch 'branch-A-Varargs'

jetlfj committed:
Use Assertions in TaskList methods

There are several methods that may fail if the parameters provided to
them are invalid.

Using Assertions to define assumptions about the program state enables
the runtime to verify them. This helps to detect mistakes in the code.

Let's document important assertions that should hold at various points
in the code.

jetlfj committed:
Improve Code Quality

The code has unnecessary this keywords and some large sections of code
grouped together.

Improving code quality makes it easier for other developers to read,
understand and modify later on.

Let's,
* remove unnecessary this keywords
* structure code logically with the use of blank lines to group related
statements together

jetlfj committed:
Merge pull request #1 from jetlfj/branch-A-Assertions

Use Assertions in TaskList methods

jetlfj committed:
Merge branch 'master' into branch-A-CodeQuality

jetlfj committed:
Merge pull request #2 from jetlfj/branch-A-CodeQuality

Improve Code Quality

jetlfj committed:
Set up CI

Puakii commented on src/test/java/EventTest.java:
```diff
@@ -0,0 +1,43 @@
+    public void addEventTest() {
+        Event event = new Event("Tutorial 1", LocalDate.parse("2022-08-25"));
+        assertEquals("[E][ ] Tutorial 1 (at: Aug 25 2022)", event.toString());
+    }
```
Should the test method name be "whatIsBeingTested_descriptionOfTestInputs_expectedOutcome" format instead? See W3.7d in CS2103T website.

Puakii commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,45 @@
+     * @return The response from Duke.
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
Perhaps you should add throws DukeException after the method?

Puakii commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,103 @@
+
+        try {
+            if (commandSplit.length == 1) {
+                switch (c) {
+                case LIST:
+                    return new ListCommand();
+                case MARK:
+                    throw new EmptyIndexException("mark");
+                case UNMARK:
+                    throw new EmptyIndexException("unmark");
+                case TODO:
+                    throw new EmptyDescriptionException("a todo");
+                case DEADLINE:
+                    throw new EmptyDescriptionException("a deadline");
+                case EVENT:
+                    throw new EmptyDescriptionException("an event");
+                case DELETE:
+                    throw new EmptyIndexException("delete");
+                case FIND:
+                    throw new DukeException("Please provide a keyword to find.");
+                default:
+                    return new ByeCommand();
```
Good correct indentation of switch statement!

Puakii reviewed (commented):
Overall, your code is very well organised with adherence to Java coding standards! However, there are some nitpicks that I have to make your code cleaner and easier to understand.

jetlfj committed:
Fix CI

jetlfj committed:
Add ability to prioritise tasks

All tasks have the same level of priority.

Allowing users to set certain tasks to high priority helps them to
identify at a glance which tasks they should complete first.

Let's,
* add a Priority field to tasks
* allow users to set tasks to high priority
* display an indicator if a task is high priority

jetlfj committed:
Merge branch 'branch-C-Priority'

jetlfj committed:
Use Streams

MinHeinA reviewed (commented):
The packages are done neatly with the relevant classes placed in the same folder (package). Follows good OO practices, good naming of variables and well code is well documented. As mentioned by another student, the name of the test methods should follow "whatIsBeingTested_descriptionOfTestInputs_expectedOutcome"

boredcoco commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+ * @version CS2103T AY22/23 Sem 1
+ */
+public class DeleteCommand extends Command {
+    private int idx;
```
Use index instead of idex as a name?

boredcoco commented on src/test/java/TodoTest.java:
```diff
@@ -0,0 +1,41 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
I really like your coding style it is really concise and neat, overall great job!

boredcoco commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+    public Deadline(String description, boolean isDone, Priority priority, LocalDate by) {
+        super(description);
+        this.by = by;
+        this.isDone = isDone;
```
Good naming of boolean here!

jetlfj committed:
Improve the GUI

jetlfj committed:
Add more automated testing

jetlfj committed:
Add a User Guide

jetlfj committed:
Update the User Guide

jetlfj committed:
Fix the User Guide

jetlfj committed:
Fix the User Guide

jetlfj committed:
Fix the User Guide

jetlfj committed:
Fix the User Guide

jetlfj committed:
Edit README

jetlfj committed:
Fix GUI resizing

jetlfj committed:
Fix the GUI

jetlfj committed:
Fix images

jetlfj committed:
Fix images

jetlfj committed:
Update JavaFX

jetlfj committed:
Fix JavaFX

jetlfj committed:
Update the User Guide
